src/coprocessor/dag/rpn_expr/impl_arithmetic.rs

```diff
@@ -222,7 +320,156 @@ mod tests {
                 .unwrap();
-            assert_eq!(output, expected, "{:?}, {:?}", output, expected);
+            assert_eq!(output, expected, "lhs={:?}, rhs={:?}", lhs, rhs);
+        }
+    }
+
+    #[test]
+    fn test_mod_int() {
+        let tests = vec![
+            (Some(13), Some(11), Some(2)),
+            (Some(-13), Some(11), Some(-2)),
+            (Some(13), Some(-11), Some(2)),
+            (Some(-13), Some(-11), Some(-2)),
+            (Some(33), Some(11), Some(0)),
+            (Some(33), Some(-11), Some(0)),
+            (Some(-33), Some(-11), Some(0)),
+            (Some(-11), None, None),
+            (None, Some(-11), None),
+            (Some(11), Some(0), None),
+            (Some(-11), Some(0), None),
+            (
+                Some(std::i64::MAX),
+                Some(std::i64::MIN),
+                Some(std::i64::MAX),
+            ),
+            (Some(std::i64::MIN), Some(std::i64::MAX), Some(-1)),
+        ];
+
+        for (lhs, rhs, expected) in tests {
+            let output = RpnFnScalarEvaluator::new()
+                .push_param(lhs)
+                .push_param(rhs)
+                .evaluate(ScalarFuncSig::ModInt)
+                .unwrap();
+            assert_eq!(output, expected, "lhs={:?}, rhs={:?}", lhs, rhs);
+        }
+    }
+    #[test]
+    fn test_mod_int_unsigned() {
+        let tests = vec![
+            (
+                Some(std::u64::MAX as i64),
+                true,
+                Some(std::i64::MIN),
+                false,
+                Some(std::i64::MAX),
+            ),
+            (
+                Some(std::i64::MIN),
+                false,
+                Some(std::u64::MAX as i64),
+                true,
+                Some(std::i64::MIN),
+            ),
+        ];
+
+        for (lhs, lhs_is_unsigned, rhs, rhs_is_unsigned, expected) in tests {
+            let lhs_field_type = FieldTypeBuilder::new()
+                .tp(FieldTypeTp::LongLong)
+                .flag(if lhs_is_unsigned {
+                    FieldTypeFlag::UNSIGNED
+                } else {
+                    FieldTypeFlag::empty()
+                })
+                .build();
+            let rhs_field_type = FieldTypeBuilder::new()
+                .tp(FieldTypeTp::LongLong)
+                .flag(if rhs_is_unsigned {
+                    FieldTypeFlag::UNSIGNED
+                } else {
+                    FieldTypeFlag::empty()
+                })
+                .build();
+            let output = RpnFnScalarEvaluator::new()
+                .push_param_with_field_type(lhs, lhs_field_type)
+                .push_param_with_field_type(rhs, rhs_field_type)
+                .evaluate(ScalarFuncSig::ModInt)
+                .unwrap();
+            assert_eq!(output, expected, "lhs={:?}, rhs={:?}", lhs, rhs);
+        }
+    }
+
+    #[test]
+    fn test_mod_real() {
+        let tests = vec![
+            (Real::new(1.0).ok(), None, None),
```

Comment:
Maybe we can write Option<f64> here and convert it below in the loop to make the test vec clearer. (I can do it later)